src/main/java/io/cryostat/discovery/KubeApiDiscovery.java

```diff
@@ -0,0 +1,702 @@
+    // This ensures realm node is persisted before initializing informers
+    @Transactional
+    void onStart(@Observes @Priority(1) StartupEvent evt) {
+        if (!enabled()) {
+            return;
+        }
+
+        if (!available()) {
+            logger.errorv("{0} enabled but environment is not Kubernetes!", getClass().getName());
+            return;
+        }
+
+        DiscoveryNode universe = DiscoveryNode.getUniverse();
+        if (DiscoveryNode.getRealm(REALM).isEmpty()) {
+            DiscoveryPlugin plugin = new DiscoveryPlugin();
+            DiscoveryNode node = DiscoveryNode.environment(REALM, BaseNodeType.REALM);
+            plugin.realm = node;
+            plugin.builtin = true;
+            universe.children.add(node);
+            node.parent = universe;
+            plugin.persist();
+            universe.persist();
+        }
+
+        logger.infov("Starting {0} client", REALM);
+    }
+
+    @Transactional
+    void onAfterStart(@Observes StartupEvent evt) {
+        safeGetInformers();
+    }
+
+    void onStop(@Observes ShutdownEvent evt) {
+        if (!(enabled() && available())) {
+            return;
+        }
+
+        logger.infov("Shutting down {0} client", REALM);
+        safeGetInformers()
+                .forEach(
+                        (ns, informer) -> {
+                            informer.close();
+                            logger.infov(
+                                    "Closed Endpoints SharedInformer for namespace \"{0}\"", ns);
+                        });
+    }
+
+    boolean enabled() {
+        return enabled;
+    }
+
+    boolean available() {
+        try {
+            boolean hasNamespace = StringUtils.isNotBlank(kubeConfig.getOwnNamespace());
+            return kubeConfig.kubeApiAvailable() && hasNamespace;
+        } catch (Exception e) {
+            logger.info(e);
+        }
+        return false;
+    }
+
+    KubernetesClient client() {
+        return kubeConfig.kubeClient();
+    }
+
+    private boolean isCompatiblePort(EndpointPort port) {
+        return jmxPortNames.orElse(List.of()).contains(port.getName())
+                || jmxPortNumbers.orElse(List.of()).contains(port.getPort());
+    }
+
+    private List<TargetTuple> getTargetTuplesFrom(Endpoints endpoints) {
+        return tuplesFromEndpoints(endpoints).stream()
+                .filter(
+                        (ref) -> {
+                            return Objects.nonNull(ref) && isCompatiblePort(ref.port);
+                        })
+                .collect(Collectors.toList());
+    }
+
+    private Map<String, SharedIndexInformer<Endpoints>> safeGetInformers() {
+        Map<String, SharedIndexInformer<Endpoints>> informers;
+        try {
+            informers = nsInformers.get();
+        } catch (ConcurrentException e) {
+            throw new IllegalStateException(e);
+        }
+        return informers;
+    }
+
+    @Transactional
```

Comment:
I don't think this annotation will work as intended, since this method is not an "entrypoint" that the Quarkus framework calls into but rather one that is invoked from other application code.

https://quarkus.io/guides/transaction#programmatic-approach

I think this example is probably suitable:

```java
        QuarkusTransaction.joiningExisting().run(() -> {
            //do work
        });
```

Reply:
No, I think that one is also just buggy :-)

Reply:
Ahh okayy haha! I will update that!